src/Rules/Actions.hs

```diff
@@ -126,6 +128,18 @@ copyDirectory source target = do
     putProgressInfo $ renderAction "Copy directory" source target
     quietly $ cmd cmdEcho ["cp", "-r", source, target]
 
+-- | Copy the content of the source directory into the target directory. Only
+-- the files and directories for which the predicate returns True are copied.
+copyDirectoryContent :: (FilePath -> IO Bool) -> FilePath -> FilePath -> Action ()
+copyDirectoryContent test source target = do
+    putProgressInfo $ renderAction "Copy directory" source target
+    liftIO $ X.listFilesInside test' source >>= mapM_ cp
+  where
+    target' a = target -/- fromJust (stripPrefix source a)
+    test' a = ifM (test a) (mkdir a >> return True) (return False)
+    mkdir a = IO.createDirectoryIfMissing True $ target' a
+    cp a = whenM (test a) $ IO.copyFile a $ target' a
```

Comment:
@KaiHa Consider the following scenario:
```bash
$ build sdist # This builds a source distribution tarball
$ vi libraries/base/Prelude.hs # Edit sources
$ build sdist # Rebuild the source distribution tarball
```
Presumably, in the second invocation of `build sdist` you expect the tarball to be updated automatically, and that nothing should be done if `libraries/base/Prelude.hs` was actually unchanged. Always removing the tarball may consume more time than necessary. It doesn't matter whether the rule is _phony_, Hadrian can still provide certain correctness and performance guarantees, e.g. _Running `build sdist` guarantees that the source tarball is up-to-date and we only rebuilt what was necessary_. This is not that hard to achieve with Shake: just don't forget to track dependencies with `need` and friends.